### ISSUE TITLE
New onig_sys is not compatible with onig-4.3.2

### ISSUE THREAD

**Issue by kornelski** [rust-onig/rust-onig#111]
There are minor differences from bindings:

```
   Compiling onig v4.3.2
error[E0412]: cannot find type `OnigRegexMut` in module `onig_sys`
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/lib.rs:142:20
    |
142 |     raw: onig_sys::OnigRegexMut,
    |                    ^^^^^^^^^^^^ help: a type alias with a similar name exists: `OnigRegex`

error[E0412]: cannot find type `OnigRegexMut` in module `onig_sys`
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/lib.rs:320:32
    |
320 |         let mut reg: onig_sys::OnigRegexMut = null_mut();
    |                                ^^^^^^^^^^^^ help: a type alias with a similar name exists: `OnigRegex`

error[E0412]: cannot find type `OnigRegexMut` in module `onig_sys`
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/lib.rs:321:50
    |
321 |         let reg_ptr = &mut reg as *mut onig_sys::OnigRegexMut;
    |                                                  ^^^^^^^^^^^^ help: a type alias with a similar name exists: `OnigRegex`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/lib.rs:326:18
    |
326 |             enc: null(),
    |                  ^^^^^^ types differ in mutability
    |
    = note: expected type `*mut onig_sys::OnigEncodingTypeST`
               found type `*const _`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/lib.rs:327:18
    |
327 |             par: null(),
    |                  ^^^^^^ types differ in mutability
    |
    = note: expected type `*mut u8`
               found type `*const _`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/lib.rs:328:22
    |
328 |             par_end: null(),
    |                      ^^^^^^ types differ in mutability
    |
    = note: expected type `*mut u8`
               found type `*const _`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/lib.rs:341:17
    |
341 |                 syntax as *const Syntax as *const onig_sys::OnigSyntaxType,
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ in mutability
    |
    = note: expected type `*mut onig_sys::OnigSyntaxType`
               found type `*const onig_sys::OnigSyntaxType`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/lib.rs:346:19
    |
346 |         if err == onig_sys::ONIG_NORMAL {
    |                   ^^^^^^^^^^^^^^^^^^^^^ expected i32, found u32
help: you can convert an `u32` to `i32` and panic if the converted value wouldn't fit
    |
346 |         if err == onig_sys::ONIG_NORMAL.try_into().unwrap() {
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/lib.rs:515:17
    |
515 |                 match_param.as_raw()
    |                 ^^^^^^^^^^^^^^^^^^^^ types differ in mutability
    |
    = note: expected type `*mut onig_sys::OnigMatchParamStruct`
               found type `*const onig_sys::OnigMatchParamStruct`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/lib.rs:704:17
    |
704 |                 match_param.as_raw()
    |                 ^^^^^^^^^^^^^^^^^^^^ types differ in mutability
    |
    = note: expected type `*mut onig_sys::OnigMatchParamStruct`
               found type `*const onig_sys::OnigMatchParamStruct`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/find.rs:182:17
    |
182 |                 scan_cb::<F>,
    |                 ^^^^^^^^^^^^ expected enum `std::option::Option`, found fn item
    |
    = note: expected type `std::option::Option<unsafe extern "C" fn(i32, i32, *mut onig_sys::re_registers, *mut std::ffi::c_void) -> i32>`
               found type `extern "C" fn(i32, i32, *const onig_sys::re_registers, *mut std::ffi::c_void) -> i32 {find::<impl Regex>::scan_with_region::scan_cb::<F>}`

error[E0308]: mismatched types
  --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/region.rs:24:22
   |
24 |                 beg: null(),
   |                      ^^^^^^ types differ in mutability
   |
   = note: expected type `*mut i32`
              found type `*const _`

error[E0308]: mismatched types
  --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/region.rs:25:22
   |
25 |                 end: null(),
   |                      ^^^^^^ types differ in mutability
   |
   = note: expected type `*mut i32`
              found type `*const _`

error[E0308]: mismatched types
  --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/region.rs:26:31
   |
26 |                 history_root: null(),
   |                               ^^^^^^ types differ in mutability
   |
   = note: expected type `*mut onig_sys::OnigCaptureTreeNodeStruct`
              found type `*const _`

error[E0308]: mismatched types
  --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/region.rs:53:57
   |
53 |             onig_sys::onig_region_copy(&mut region.raw, ptr);
   |                                                         ^^^ types differ in mutability
   |
   = note: expected type `*mut onig_sys::re_registers`
              found type `*const onig_sys::re_registers`

error[E0308]: mismatched types
  --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/region.rs:84:17
   |
84 |         if r != onig_sys::ONIG_NORMAL {
   |                 ^^^^^^^^^^^^^^^^^^^^^ expected i32, found u32
help: you can convert an `u32` to `i32` and panic if the converted value wouldn't fit
   |
84 |         if r != onig_sys::ONIG_NORMAL.try_into().unwrap() {
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/region.rs:131:61
    |
131 |         let tree = unsafe { onig_sys::onig_get_capture_tree(&self.raw) };
    |                                                             ^^^^^^^^^ types differ in mutability
    |
    = note: expected type `*mut onig_sys::re_registers`
               found type `&onig_sys::re_registers`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/region.rs:191:17
    |
191 |                 &self.raw,
    |                 ^^^^^^^^^ types differ in mutability
    |
    = note: expected type `*mut onig_sys::re_registers`
               found type `&onig_sys::re_registers`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/region.rs:193:17
    |
193 |                 traverse_cb::<F>,
    |                 ^^^^^^^^^^^^^^^^
    |                 |
    |                 expected enum `std::option::Option`, found fn item
    |                 help: try using a variant of the expected type: `Some(traverse_cb::<F>)`
    |
    = note: expected type `std::option::Option<unsafe extern "C" fn(i32, i32, i32, i32, i32, *mut std::ffi::c_void) -> i32>`
               found type `extern "C" fn(i32, i32, i32, i32, i32, *mut std::ffi::c_void) -> i32 {region::Region::tree_traverse_at::traverse_cb::<F>}`

error[E0308]: mismatched types
  --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/match_param.rs:30:17
   |
30 |                 limit as c_uint
   |                 ^^^^^^^^^^^^^^^
   |                 |
   |                 expected u64, found u32
   |                 help: you can convert an `u32` to `u64`: `(limit as c_uint).into()`

error[E0308]: mismatched types
  --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/names.rs:64:17
   |
64 |                 foreach_cb::<F>,
   |                 ^^^^^^^^^^^^^^^ expected enum `std::option::Option`, found fn item
   |
   = note: expected type `std::option::Option<unsafe extern "C" fn(*const u8, *const u8, i32, *mut i32, *mut onig_sys::re_pattern_buffer, *mut std::ffi::c_void) -> i32>`
              found type `extern "C" fn(*const u8, *const u8, i32, *const i32, *mut onig_sys::re_pattern_buffer, *mut std::ffi::c_void) -> i32 {names::<impl Regex>::foreach_name::foreach_cb::<F>}`

error[E0308]: mismatched types
  --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/syntax.rs:99:51
   |
99 |             let op = onig_sys::onig_get_syntax_op(&self.raw);
   |                                                   ^^^^^^^^^ types differ in mutability
   |
   = note: expected type `*mut onig_sys::OnigSyntaxType`
              found type `&onig_sys::OnigSyntaxType`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/syntax.rs:100:53
    |
100 |             let op2 = onig_sys::onig_get_syntax_op2(&self.raw);
    |                                                     ^^^^^^^^^ types differ in mutability
    |
    = note: expected type `*mut onig_sys::OnigSyntaxType`
               found type `&onig_sys::OnigSyntaxType`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/syntax.rs:135:88
    |
135 |         SyntaxBehavior::from_bits_truncate(unsafe { onig_sys::onig_get_syntax_behavior(&self.raw) })
    |                                                                                        ^^^^^^^^^ types differ in mutability
    |
    = note: expected type `*mut onig_sys::OnigSyntaxType`
               found type `&onig_sys::OnigSyntaxType`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/syntax.rs:160:85
    |
160 |         RegexOptions::from_bits_truncate(unsafe { onig_sys::onig_get_syntax_options(&self.raw) })
    |                                                                                     ^^^^^^^^^ types differ in mutability
    |
    = note: expected type `*mut onig_sys::OnigSyntaxType`
               found type `&onig_sys::OnigSyntaxType`

error[E0308]: mismatched types
  --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/utils.rs:37:77
   |
37 |         unsafe { onig_sys::onig_unicode_define_user_property(name.as_ptr(), raw_ranges.as_ptr()) };
   |                                                                             ^^^^^^^^^^^^^^^^^^^ types differ in mutability
   |
   = note: expected type `*mut u32`
              found type `*const u32`

error[E0308]: mismatched types
  --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/buffers.rs:28:18
   |
27 |     fn encoding(&self) -> onig_sys::OnigEncoding {
   |                           ---------------------- expected `*mut onig_sys::OnigEncodingTypeST` because of return type
28 |         unsafe { &onig_sys::OnigEncodingUTF8 }
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ in mutability
   |
   = note: expected type `*mut onig_sys::OnigEncodingTypeST`
              found type `&onig_sys::OnigEncodingTypeST`

error[E0308]: mismatched types
  --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/buffers.rs:97:27
   |
97 |             enc: unsafe { &onig_sys::OnigEncodingASCII },
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ types differ in mutability
   |
   = note: expected type `*mut onig_sys::OnigEncodingTypeST`
              found type `&onig_sys::OnigEncodingTypeST`

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/flags.rs:263:13
    |
263 |             onig_sys::ONIG_TRAVERSE_CALLBACK_AT_FIRST;
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected i32, found u32

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/flags.rs:266:13
    |
266 |             onig_sys::ONIG_TRAVERSE_CALLBACK_AT_LAST;
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected i32, found u32

error[E0308]: mismatched types
   --> ~/.cargo/registry/src/github.com-1ecc6299db9ec823/onig-4.3.2/src/flags.rs:269:13
    |
269 |             onig_sys::ONIG_TRAVERSE_CALLBACK_AT_BOTH;
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected i32, found u32
```

**Comment by kornelski** [comment]
Things like `i32`/`u32` can be worked around using `as _` syntax, e.g.


     if err == onig_sys::ONIG_NORMAL as _ {

**Comment by iwillspeak** [comment]
I'm going make a new build of the 4.x version and pin the`onig` release to `6.9.1`

**Comment by iwillspeak** [comment]
Hopefully this is fixed by #112 .
